Importing all libarries

In [1]:
import numpy as np
import pandas as pd
import tensorflow as tf
import sklearn as sns

Importing our datasets

In [2]:
data = pd.read_csv("Churn_Modelling.csv")
data.head(5)
#data["Gender"].unique()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

Seperating our datasets into x and y variable

In [3]:
x = data.iloc[:,3:-1].values
y = data.iloc[:,-1].values

In [4]:
x

array([[619, 'France', 'Female', ..., 1, 1, 101348.88],
       [608, 'Spain', 'Female', ..., 0, 1, 112542.58],
       [502, 'France', 'Female', ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 'Female', ..., 0, 1, 42085.58],
       [772, 'Germany', 'Male', ..., 1, 0, 92888.52],
       [792, 'France', 'Female', ..., 1, 0, 38190.78]], dtype=object)

In [5]:
y

array([1, 0, 1, ..., 1, 1, 0], dtype=int64)

Label encoding in gender column

In [6]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
x[:,2] = le.fit_transform(x[:,2])
x

array([[619, 'France', 0, ..., 1, 1, 101348.88],
       [608, 'Spain', 0, ..., 0, 1, 112542.58],
       [502, 'France', 0, ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 0, ..., 0, 1, 42085.58],
       [772, 'Germany', 1, ..., 1, 0, 92888.52],
       [792, 'France', 0, ..., 1, 0, 38190.78]], dtype=object)

Checking unique values 

In [7]:
data["Geography"].unique()

array(['France', 'Spain', 'Germany'], dtype=object)

Onehot encoding in Geography column

In [8]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
ct = ColumnTransformer(transformers=[('encoder',OneHotEncoder(),[1])],remainder='passthrough')
x = np.array(ct.fit_transform(x))
print(x)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


Spliting datasets into training and testing 

In [9]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=0)

In [10]:
x_train.shape

(8000, 12)

In [11]:
y_train.shape

(8000,)

In [12]:
x_test.shape

(2000, 12)

In [13]:
y_test.shape

(2000,)

Features scaling

In [14]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.fit_transform(x_test)

In [15]:
x_train

array([[-1.01460667, -0.5698444 ,  1.74309049, ...,  0.64259497,
        -1.03227043,  1.10643166],
       [-1.01460667,  1.75486502, -0.57369368, ...,  0.64259497,
         0.9687384 , -0.74866447],
       [ 0.98560362, -0.5698444 , -0.57369368, ...,  0.64259497,
        -1.03227043,  1.48533467],
       ...,
       [ 0.98560362, -0.5698444 , -0.57369368, ...,  0.64259497,
        -1.03227043,  1.41231994],
       [-1.01460667, -0.5698444 ,  1.74309049, ...,  0.64259497,
         0.9687384 ,  0.84432121],
       [-1.01460667,  1.75486502, -0.57369368, ...,  0.64259497,
        -1.03227043,  0.32472465]])

In [16]:
x_test

array([[-0.95692675,  1.62776996, -0.57427105, ...,  0.66011376,
         0.97628121,  1.62185911],
       [ 1.04501206, -0.61433742, -0.57427105, ...,  0.66011376,
        -1.02429504,  0.504204  ],
       [-0.95692675, -0.61433742,  1.74133801, ...,  0.66011376,
         0.97628121, -0.41865644],
       ...,
       [-0.95692675, -0.61433742,  1.74133801, ...,  0.66011376,
        -1.02429504,  0.72775202],
       [-0.95692675,  1.62776996, -0.57427105, ...,  0.66011376,
         0.97628121, -1.54162886],
       [-0.95692675,  1.62776996, -0.57427105, ...,  0.66011376,
        -1.02429504,  1.62356528]])

Initializing the ANN

In [17]:
ann = tf.keras.models.Sequential()

adding input and 1st layer

In [18]:
ann.add(tf.keras.layers.Dense(units=6,activation='relu'))

adding 2nd layer 

In [19]:
ann.add(tf.keras.layers.Dense(units=6,activation='relu'))

Adding output layer

In [20]:
ann.add(tf.keras.layers.Dense(units=1,activation='sigmoid'))

In [21]:
ann

Compiling the ANN

In [22]:
ann.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [23]:
ann.fit(x_train,y_train,batch_size=32,epochs=50)

Epoch 1/50
250/250 [==============================] - 2s 3ms/step - loss: 0.5118 - accuracy: 0.7960
Epoch 2/50
250/250 [==============================] - 1s 3ms/step - loss: 0.4641 - accuracy: 0.7960
Epoch 3/50
250/250 [==============================] - 1s 3ms/step - loss: 0.4440 - accuracy: 0.7960
Epoch 4/50
250/250 [==============================] - 1s 3ms/step - loss: 0.4331 - accuracy: 0.8039
Epoch 5/50
250/250 [==============================] - 1s 3ms/step - loss: 0.4266 - accuracy: 0.8144
Epoch 6/50
250/250 [==============================] - 1s 3ms/step - loss: 0.4219 - accuracy: 0.8188
Epoch 7/50
250/250 [==============================] - 1s 2ms/step - loss: 0.4186 - accuracy: 0.8246
Epoch 8/50
250/250 [==============================] - 1s 2ms/step - loss: 0.4151 - accuracy: 0.8249
Epoch 9/50
250/250 [==============================] - 1s 2ms/step - loss: 0.4113 - accuracy: 0.8281
Epoch 10/50
250/250 [==============================] - 1s 2ms/step - loss: 0.4074 - accuracy: 0.8321

Predicting the result of a single observation

Using ANN model to predict if the customer with the following informations will leave the bank:

Geography: Spain

Credit Score: 600

Gender: Female

Age: 40 years old

Tenure: 3 years

Balance: $ 60000

Number of Products: 2

Does this customer have a credit card ? Yes

Is this customer an Active Member: Yes

Estimated Salary: $ 50000

In [24]:
print(ann.predict(sc.transform([[600,0,0,1,0,40,3,60000,2,1,1,50000]]))>0.5)

1/1 [==============================] - 0s 300ms/step
[[False]]


As we can see in the above results value is less then 50% means we can say it is close to 0 so person will stay in the bank

Predicting the Test set results

In [25]:
y_pred = ann.predict(x_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

63/63 [==============================] - 0s 3ms/step
[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


Checking accuracy

In [26]:
from sklearn.metrics import accuracy_score,confusion_matrix
cm = confusion_matrix(y_test,y_pred)
print(cm)

[[1540   55]
 [ 224  181]]


In [27]:
accuracy_score(y_pred,y_test)

0.8605

As we are geeting 86% accuracy so we can use this model